solidity/contracts/Rewards.sol

```diff
@@ -540,6 +551,55 @@ contract Rewards is Ownable {
         }
     }
 
+    /// @notice Initiates the process of upgrading to another rewards contract.
+    /// @param _newRewardsContract The address of a new rewards contract.
+    function initiateRewardsUpgrade(address _newRewardsContract) public onlyOwner {
+        upgradeInitiatedTimestamp = block.timestamp;
+        newRewardsContract = _newRewardsContract;
+        emit UpgradeInitiated(newRewardsContract);
+    }
+
+    /// @notice Finalizes the process of upgrading to another rewards contract
+    /// by allocating all past intervals and then, transferring the
+    /// not-yet-allocated tokens to a new rewards contract.
+    /// Can be called only when the interval during which the upgrade was
+    /// initiated ended.
+    /// Before finalizing the upgrade, make sure all terminated groups are
+    /// reported.
+    function finalizeRewardsUpgrade() public onlyOwner {
+        require(upgradeInitiatedTimestamp != 0, "Upgrade not initiated");
+
+        // allocate all intervals that can be currently allocated
+        uint256 currentInterval = intervalOf(block.timestamp);
+        uint256 intervalToAllocate = 0;
+        if (currentInterval > 0) {
```

Comment:
My understanding of "0" intervals is that they all ended in September. Then from now on, we're dealing with intervals 1,2,3... so `currrentInterval` will always be > 0?

Reply:
There are two things to keep in mind:
- `Rewards` contract is abstract and it does not know when the first interval starts and ends,
- We are allocating previous intervals here - it does not matter if the interval ended but whether it was allocated.

Reply:
Okay, makes sense.